BUG: pd.testing.assert_frame_equal(..., check_exact=True) raises AssertionError when comparing numeric ExtensionDtypes

- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd

d = {"x": [1, 2]}

df1 = pd.DataFrame(data=d, dtype=pd.UInt32Dtype())
df2 = pd.DataFrame(data=d, dtype=pd.UInt32Dtype())

pd.testing.assert_frame_equal(df1, df2, check_exact=True)
```

#### Problem description

The above code raises the following exception:

```
AssertionError: ndarray Expected type <class 'numpy.ndarray'>, found <class 'pandas.core.arrays.integer.IntegerArray'> instead
```

Instead, the usage of `pd.testing.assert_frame_equal(..., check_exact=True)` should not raise an exception when comparing numeric `ExtensionDtype`s.

#### Expected Output

No exception.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 9d598a5e1eee26df95b3910e3f2934890d062caa
python           : 3.8.7.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 17.7.0
Version          : Darwin Kernel Version 17.7.0: Fri Oct 30 13:34:27 PDT 2020; root:xnu-4570.71.82.8~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.1
numpy            : 1.19.5
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 20.2.3
setuptools       : 49.2.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

--- a/doc/source/whatsnew/v1.2.2.rst
+++ b/doc/source/whatsnew/v1.2.2.rst
@@ -14,7 +14,7 @@ including other versions of pandas.
 
 Fixed regressions
 ~~~~~~~~~~~~~~~~~
--
+- Fixed regression in :func:`pandas.testing.assert_series_equal` and :func:`pandas.testing.assert_frame_equal` always raising ``AssertionError`` when comparing extension dtypes (:issue:`39410`)
 -
 
 .. ---------------------------------------------------------------------------

--- a/pandas/_testing/asserters.py
+++ b/pandas/_testing/asserters.py
@@ -968,14 +968,24 @@ def assert_series_equal(
             assert_attr_equal("dtype", left, right, obj=f"Attributes of {obj}")
 
     if check_exact and is_numeric_dtype(left.dtype) and is_numeric_dtype(right.dtype):
+        left_values = left._values
+        right_values = right._values
         # Only check exact if dtype is numeric
-        assert_numpy_array_equal(
-            left._values,
-            right._values,
-            check_dtype=check_dtype,
-            obj=str(obj),
-            index_values=np.asarray(left.index),
-        )
+        if is_extension_array_dtype(left_values):
+            assert_extension_array_equal(
+                left_values,
+                right_values,
+                check_dtype=check_dtype,
+                index_values=np.asarray(left.index),
+            )
+        else:
+            assert_numpy_array_equal(
+                left_values,
+                right_values,
+                check_dtype=check_dtype,
+                obj=str(obj),
+                index_values=np.asarray(left.index),
+            )
     elif check_datetimelike_compat and (
         needs_i8_conversion(left.dtype) or needs_i8_conversion(right.dtype)
     ):

--- a/pandas/_testing/asserters.py
+++ b/pandas/_testing/asserters.py
@@ -971,7 +971,9 @@ def assert_series_equal(
         left_values = left._values
         right_values = right._values
         # Only check exact if dtype is numeric
-        if is_extension_array_dtype(left_values):
+        if is_extension_array_dtype(left_values) and is_extension_array_dtype(
+            right_values
+        ):
             assert_extension_array_equal(
                 left_values,
                 right_values,

--- a/pandas/tests/util/test_assert_frame_equal.py
+++ b/pandas/tests/util/test_assert_frame_equal.py
@@ -305,3 +305,19 @@ def test_assert_frame_equal_columns_mixed_dtype():
     # GH#39168
     df = DataFrame([[0, 1, 2]], columns=["foo", "bar", 42], index=[1, "test", 2])
     tm.assert_frame_equal(df, df, check_like=True)
+
+
+def test_frame_equal_extension_dtype(frame_or_series, any_nullable_numeric_dtype):
+    # GH#39410
+    obj = frame_or_series([1, 2], dtype=any_nullable_numeric_dtype)
+    tm.assert_equal(obj, obj, check_exact=True)
+
+
+@pytest.mark.parametrize("indexer", [(0, 1), (1, 0)])
+def test_frame_equal_mixed_dtypes(frame_or_series, any_nullable_numeric_dtype, indexer):
+    dtypes = (any_nullable_numeric_dtype, "int64")
+    obj1 = frame_or_series([1, 2], dtype=dtypes[indexer[0]])
+    obj2 = frame_or_series([1, 2], dtype=dtypes[indexer[1]])
+    msg = r'(Series|DataFrame.iloc\[:, 0\] \(column name="0"\) classes) are different'
+    with pytest.raises(AssertionError, match=msg):
+        tm.assert_equal(obj1, obj2, check_exact=True, check_dtype=False)

